sync3/connmap_test.go

```diff
@@ -0,0 +1,211 @@
+		otherCID: nil,
+	}
+	for cid := range cidToConn {
+		_, cancel := context.WithCancel(context.Background())
+		conn := cm.CreateConn(cid, cancel, func() ConnHandler {
+			return &mockConnHandler{}
+		})
+		cidToConn[cid] = conn
+	}
+
+	closedDevice := "A"
+	cm.CloseConnsForDevice(alice, closedDevice)
+	time.Sleep(100 * time.Millisecond) // some stuff happens asyncly in goroutines
+
+	// Destroy should have been called for all alice|A connections
+	assertDestroyedConns(t, cidToConn, func(cid ConnID) bool {
+		return cid.UserID == alice && cid.DeviceID == "A"
+	})
+}
+
+func TestConnMap_CloseConnsForUser(t *testing.T) {
+	cm := NewConnMap(false, time.Minute)
+	otherCID := ConnID{UserID: bob, DeviceID: "A", CID: "room-list"}
+	cidToConn := map[ConnID]*Conn{
+		{UserID: alice, DeviceID: "A", CID: "room-list"}:     nil,
+		{UserID: alice, DeviceID: "A", CID: "encryption"}:    nil,
+		{UserID: alice, DeviceID: "A", CID: "notifications"}: nil,
+		{UserID: alice, DeviceID: "B", CID: "room-list"}:     nil,
+		{UserID: alice, DeviceID: "B", CID: "encryption"}:    nil,
+		{UserID: alice, DeviceID: "B", CID: "notifications"}: nil,
+		otherCID: nil,
+	}
+	for cid := range cidToConn {
+		_, cancel := context.WithCancel(context.Background())
+		conn := cm.CreateConn(cid, cancel, func() ConnHandler {
+			return &mockConnHandler{}
+		})
+		cidToConn[cid] = conn
+	}
+
+	num := cm.CloseConnsForUsers([]string{alice})
+	time.Sleep(100 * time.Millisecond) // some stuff happens asyncly in goroutines
+	must.Equal(t, num, 6, "unexpected number of closed conns")
+
+	// Destroy should have been called for all alice|A connections
+	assertDestroyedConns(t, cidToConn, func(cid ConnID) bool {
+		return cid.UserID == alice
+	})
+}
+
+func TestConnMap_TTLExpiry(t *testing.T) {
+	cm := NewConnMap(false, time.Second) // 1s expiry
+	expiredCIDs := []ConnID{
+		{UserID: alice, DeviceID: "A", CID: "room-list"},
+		{UserID: alice, DeviceID: "A", CID: "encryption"},
+		{UserID: alice, DeviceID: "A", CID: "notifications"},
+	}
+	cidToConn := map[ConnID]*Conn{}
+	for _, cid := range expiredCIDs {
+		_, cancel := context.WithCancel(context.Background())
+		conn := cm.CreateConn(cid, cancel, func() ConnHandler {
+			return &mockConnHandler{}
+		})
+		cidToConn[cid] = conn
+	}
+	time.Sleep(time.Millisecond * 500)
+
+	unexpiredCIDs := []ConnID{
+		{UserID: alice, DeviceID: "B", CID: "room-list"},
+		{UserID: alice, DeviceID: "B", CID: "encryption"},
+		{UserID: alice, DeviceID: "B", CID: "notifications"},
+	}
+	for _, cid := range unexpiredCIDs {
+		_, cancel := context.WithCancel(context.Background())
+		conn := cm.CreateConn(cid, cancel, func() ConnHandler {
+			return &mockConnHandler{}
+		})
+		cidToConn[cid] = conn
+	}
+
+	time.Sleep(510 * time.Millisecond) // all 'A' device conns must have expired
+
+	// Destroy should have been called for all alice|A connections
+	assertDestroyedConns(t, cidToConn, func(cid ConnID) bool {
+		return cid.DeviceID == "A"
```

Comment:
should this test also include a device from bob and then check the user id here?